### PR TITLE
Adjust the 50 move rule

### PR DESCRIPTION
**Pull request by MelleKoning**
Introducing --draw-move-rule parameter.

For analyse purposes it is interesting to adjust the 50 move rule, to see if lc0 chooses 'more decisive' moves when lowering this number. 

The 50 move rule draws the game when no pieces and no pawn moves have been done in 50 moves or 100 ply's. With this param this number can be adjusted.

To illustrate, an example from TCEC game 82.

https://cd.tcecbeta.club/archive.html?season=15&div=sf&game=82#/

image

At move 60 StockFish moves away the rook b7-a7, Leela of course moved up the gap with Rc8-b8, to go for the B-pawn capture.

However, after StockFish played Bg7-f8, Leela awaits the attack with a few typical waiting moves like knight moves (Nb3-d2, Ka2-b2, Nd2-b1, Nb1-d2 and yes Nd2 back to b3) before finally capturing the pawn with Rb8-b6!

This PR is to inspect if lowering the draw-move-rule to (say) 10 influences the choice of leela.
This indeed turns out to be the case (not too surprising of course, but interesting for analysis nevertheless).

As tested first with default:
```
lc0 --verbose-move-stats -t 1 --minibatch-size=16 --weights=.\nets\42372  
       _
|   _ | |
|_ |_ |_| v0.22.0-dev built May 17 2019
position startpos moves d2d4 g8f6 c2c4 g7g6 b1c3 f8g7 e2e4 d7d6 f2f3 e8g8 c1e3 e7e5 d4d5 c7c6 d1d2 c6d5 c4d5 b8a6 g2g4 a6c5 h2h4 h7h5 g4g5 f6e8 b2b4 c5d7 g1h3 a7a5 b4b5 d7c5 h3f2 g8h7 f1e2 b7b6 f2d3 f7f5 h1f1 f8f7 e1c1 f7b7 c1b1 c8d7 d3b2 b7c7 e3c5 c7c5 d1c1 f5f4 f1d1 g7f8 e2f1 a8a7 d2c2 a7b7 c2b3 f8e7 c3a4 c5c1 d1c1 e8c7 c1c6 c7a8 a4c3 h7g7 b2c4 d7e8 f1h3 b7b8 b3d1 a8c7 a2a4 e7f8 d1d3 g7h7 c4b2 h7g7 c6c4 d8e7 d3d2 g7h7 c3e2 f8g7 c4c3 c7a8 e2c1 b8b7 c3c8 a8c7 d2c2 g7f8 c1b3 h7g8 c2c4 g8g7 b2d3 g7h7 b3d2 f8g7 b1a1 g7f8 a1a2 h7g8 a2b1 g8h7 b1a1 f8g7 a1a2 g7f8 d2b1 e7f7 b1a3 f8g7 d3c1 f7e7 c1b3 e8f7 a3b1 f7e8 b1c3 b7a7 c8b8 g7f8
go nodes 120000
info depth 8 seldepth 17 time 58830 nodes 119950 score cp 1606 hashfull 636 nps 361 tbhits 0 pv b3d2 c7a8 c4c8 a7c7 c8e8 c7c3 e8e7 f8e7 a2b2 c3d3 b2c2 d3a3 b8a8 a3a2 c2c3 a2a3
info string c4c7  (735 ) N:       5 (+ 0) (P:  0.46%) (Q: -0.75445) (D:  0.000) (U: 1.82908) (Q+U:  1.07462) (V: -0.8310)
info string c4d4  (720 ) N:       6 (+ 0) (P:  0.56%) (Q: -0.91347) (D:  0.000) (U: 1.91705) (Q+U:  1.00358) (V: -0.5405)
info string b8a8  (1638) N:       6 (+ 0) (P:  0.56%) (Q: -0.85634) (D:  0.000) (U: 1.93601) (Q+U:  1.07968) (V: -0.7163)
info string c4c5  (727 ) N:       6 (+ 0) (P:  0.56%) (Q: -0.80621) (D:  0.000) (U: 1.90724) (Q+U:  1.10102) (V: -0.9183)
info string c4b4  (719 ) N:       7 (+ 0) (P:  0.72%) (Q: -0.99052) (D:  0.000) (U: 2.17475) (Q+U:  1.18423) (V: -0.9559)
info string b8d8  (1640) N:       7 (+ 0) (P:  0.62%) (Q: -0.56535) (D:  0.000) (U: 1.84911) (Q+U:  1.28376) (V:  0.3291)
info string c4c6  (732 ) N:       8 (+ 0) (P:  0.67%) (Q: -0.59898) (D:  0.000) (U: 1.79219) (Q+U:  1.19322) (V:  0.7136)
info string h3g4  (641 ) N:       9 (+ 0) (P:  0.70%) (Q: -0.65100) (D:  0.000) (U: 1.68669) (Q+U:  1.03569) (V: -0.3704)
info string b8b7  (1635) N:       9 (+ 0) (P:  0.84%) (Q: -0.78838) (D:  0.000) (U: 2.01725) (Q+U:  1.22886) (V:  0.1662)
info string h3d7  (648 ) N:      11 (+ 0) (P:  0.65%) (Q: -0.21356) (D:  0.000) (U: 1.30599) (Q+U:  1.09244) (V: -0.0172)
info string b8e8  (1641) N:      13 (+ 0) (P:  0.65%) (Q: -0.04510) (D:  0.000) (U: 1.12138) (Q+U:  1.07628) (V:  0.1573)
info string b3d4  (455 ) N:      13 (+ 0) (P:  0.83%) (Q: -0.30291) (D:  0.000) (U: 1.42650) (Q+U:  1.12359) (V:  0.5983)
info string b3c5  (458 ) N:      21 (+ 0) (P:  0.86%) (Q:  0.14218) (D:  0.000) (U: 0.93982) (Q+U:  1.08200) (V:  0.7263)
info string h3c8  (650 ) N:      77 (+ 0) (P:  2.45%) (Q:  0.35180) (D:  0.000) (U: 0.75368) (Q+U:  1.10548) (V:  0.9635)
info string h3f5  (643 ) N:      95 (+ 0) (P:  1.00%) (Q:  0.77230) (D:  0.000) (U: 0.24905) (Q+U:  1.02134) (V:  0.8187)
info string b8c8  (1639) N:     209 (+ 0) (P:  1.20%) (Q:  0.87160) (D:  0.000) (U: 0.13744) (Q+U:  1.00904) (V:  0.8090)
info string b3a5  (456 ) N:     284 (+ 0) (P:  1.26%) (Q:  0.89180) (D:  0.000) (U: 0.10612) (Q+U:  0.99792) (V:  0.9524)
info string h3g2  (631 ) N:     368 (+ 0) (P:  1.61%) (Q:  0.89517) (D:  0.000) (U: 0.10477) (Q+U:  0.99994) (V:  0.9370)
info string c4f1  (707 ) N:     522 (+ 0) (P:  1.73%) (Q:  0.91395) (D:  0.000) (U: 0.07963) (Q+U:  0.99358) (V:  0.9565)
info string h3f1  (627 ) N:     645 (+ 0) (P:  1.69%) (Q:  0.92652) (D:  0.000) (U: 0.06277) (Q+U:  0.98929) (V:  0.9492)
info string c4d3  (716 ) N:     693 (+ 0) (P:  1.80%) (Q:  0.92670) (D:  0.000) (U: 0.06241) (Q+U:  0.98911) (V:  0.9439)
info string c4e2  (712 ) N:     760 (+ 0) (P:  2.35%) (Q:  0.91793) (D:  0.000) (U: 0.07412) (Q+U:  0.99205) (V:  0.9543)
info string a2a3  (204 ) N:     844 (+ 0) (P:  1.95%) (Q:  0.93067) (D:  0.000) (U: 0.05546) (Q+U:  0.98613) (V:  0.9673)
info string c3e2  (475 ) N:    1091 (+ 0) (P:  2.02%) (Q:  0.93987) (D:  0.000) (U: 0.04431) (Q+U:  0.98418) (V:  0.9766)
info string b3c1  (439 ) N:    1126 (+ 0) (P:  2.30%) (Q:  0.93612) (D:  0.000) (U: 0.04903) (Q+U:  0.98515) (V:  0.9654)
info string b3a1  (437 ) N:    1128 (+ 0) (P:  1.98%) (Q:  0.94157) (D:  0.000) (U: 0.04201) (Q+U:  0.98358) (V:  0.9621)
info string a2a1  (194 ) N:    1685 (+ 0) (P:  1.78%) (Q:  0.95381) (D:  0.000) (U: 0.02530) (Q+U:  0.97910) (V:  0.9696)
info string a2b1  (195 ) N:    2895 (+ 0) (P:  2.85%) (Q:  0.95439) (D:  0.000) (U: 0.02364) (Q+U:  0.97803) (V:  0.9712)
info string c3d1  (469 ) N:    3286 (+ 0) (P:  4.65%) (Q:  0.94663) (D:  0.000) (U: 0.03393) (Q+U:  0.98056) (V:  0.9805)
info string h3e6  (646 ) N:    7295 (+ 0) (P:  8.68%) (Q:  0.94812) (D:  0.000) (U: 0.02854) (Q+U:  0.97666) (V:  0.9736)
info string b8b6  (1631) N:   12627 (+ 0) (P: 16.88%) (Q:  0.94149) (D:  0.000) (U: 0.03208) (Q+U:  0.97357) (V:  0.9700)
info string c3b1  (467 ) N:   16927 (+ 0) (P: 10.75%) (Q:  0.95612) (D:  0.000) (U: 0.01524) (Q+U:  0.97136) (V:  0.9794)
info string a2b2  (197 ) N:   33598 (+ 0) (P: 10.86%) (Q:  0.96158) (D:  0.000) (U: 0.00776) (Q+U:  0.96934) (V:  0.9756)
info string b3d2  (444 ) N:   33673 (+16) (P: 11.53%) (Q:  0.96113) (D:  0.000) (U: 0.00822) (Q+U:  0.96935) (V:  0.9816)
bestmove b3d2 ponder c7a8
```

Interestingly, although leela thinks about Rb8xb6, eventually after a long time she chooses the waiting move with the Knight b3-b2, unfortunately.

Now with this PR, starting deliberately with only 10 ply of draw-move-rule, sticky-games hits in quickly.

```
lc0moverule --draw-move-rule=10 --verbose-move-stats -t 2 --minibatch-size=256 --weights=.\nets\42372
position startpos moves d2d4 g8f6 c2c4 g7g6 b1c3 f8g7 e2e4 d7d6 f2f3 e8g8 c1e3 e7e5 d4d5 c7c6 d1d2 c6d5 c4d5 b8a6 g2g4 a6c5 h2h4 h7h5 g4g5 f6e8 b2b4 c5d7 g1h3 a7a5 b4b5 d7c5 h3f2 g8h7 f1e2 b7b6 f2d3 f7f5 h1f1 f8f7 e1c1 f7b7 c1b1 c8d7 d3b2 b7c7 e3c5 c7c5 d1c1 f5f4 f1d1 g7f8 e2f1 a8a7 d2c2 a7b7 c2b3 f8e7 c3a4 c5c1 d1c1 e8c7 c1c6 c7a8 a4c3 h7g7 b2c4 d7e8 f1h3 b7b8 b3d1 a8c7 a2a4 e7f8 d1d3 g7h7 c4b2 h7g7 c6c4 d8e7 d3d2 g7h7 c3e2 f8g7 c4c3 c7a8 e2c1 b8b7 c3c8 a8c7 d2c2 g7f8 c1b3 h7g8 c2c4 g8g7 b2d3 g7h7 b3d2 f8g7 b1a1 g7f8 a1a2 h7g8 a2b1 g8h7 b1a1 f8g7 a1a2 g7f8 d2b1 e7f7 b1a3 f8g7 d3c1 f7e7 c1b3 e8f7 a3b1 f7e8 b1c3 b7a7 c8b8 g7f8
go nodes 100000
info depth 9 seldepth 21 time 49060 nodes 51258 score cp 881 hashfull 175 nps 932 tbhits 0 pv b8b6 e7d8 b6c6 c7a8 c6c8 d8e7 c8b8 a7c7 c4d3 c7a7 b5b6 a7b7 b8a8 b7b6 b3a5 b6b2 a2b2 e7g7
info string c4c7  (735 ) N:       3 (+ 0) (P:  0.46%) (Q: -0.78194) (D:  0.000) (U: 1.44288) (Q+U:  0.66094) (V: -0.8310)
info string c4c5  (727 ) N:       7 (+ 0) (P:  0.56%) (Q:  0.00000) (D:  1.000) (U: 0.87765) (Q+U:  0.87765) (V:  0.0000) (T)
info string c4d4  (720 ) N:       7 (+ 0) (P:  0.56%) (Q:  0.00000) (D:  1.000) (U: 0.88216) (Q+U:  0.88216) (V:  0.0000) (T)
info string b8a8  (1638) N:       7 (+ 0) (P:  0.56%) (Q:  0.00000) (D:  1.000) (U: 0.89089) (Q+U:  0.89089) (V:  0.0000) (T)
info string b8d8  (1640) N:       8 (+ 0) (P:  0.62%) (Q:  0.00000) (D:  1.000) (U: 0.86440) (Q+U:  0.86440) (V:  0.0000) (T)
info string b8e8  (1641) N:       8 (+ 0) (P:  0.65%) (Q: -0.01370) (D:  0.000) (U: 0.91737) (Q+U:  0.90367) (V:  0.1573)
info string h3d7  (648 ) N:       8 (+ 0) (P:  0.65%) (Q:  0.00000) (D:  1.000) (U: 0.91577) (Q+U:  0.91577) (V:  0.0000) (T)
info string c4c6  (732 ) N:       9 (+ 0) (P:  0.67%) (Q:  0.00000) (D:  1.000) (U: 0.84827) (Q+U:  0.84827) (V:  0.0000) (T)
info string h3g4  (641 ) N:       9 (+ 0) (P:  0.70%) (Q:  0.00000) (D:  1.000) (U: 0.88704) (Q+U:  0.88704) (V:  0.0000) (T)
info string c4b4  (719 ) N:       9 (+ 0) (P:  0.72%) (Q:  0.00000) (D:  1.000) (U: 0.91497) (Q+U:  0.91497) (V:  0.0000) (T)
info string b3d4  (455 ) N:      11 (+ 0) (P:  0.83%) (Q:  0.00000) (D:  1.000) (U: 0.87524) (Q+U:  0.87524) (V:  0.0000) (T)
info string b8b7  (1635) N:      11 (+ 0) (P:  0.84%) (Q:  0.00000) (D:  1.000) (U: 0.88407) (Q+U:  0.88407) (V:  0.0000) (T)
info string b3c5  (458 ) N:      11 (+ 0) (P:  0.86%) (Q:  0.00000) (D:  1.000) (U: 0.90614) (Q+U:  0.90614) (V:  0.0000) (T)
info string h3f5  (643 ) N:      13 (+ 0) (P:  1.00%) (Q:  0.00000) (D:  1.000) (U: 0.89811) (Q+U:  0.89811) (V:  0.0000) (T)
info string b8c8  (1639) N:      16 (+ 0) (P:  1.20%) (Q:  0.00000) (D:  1.000) (U: 0.89287) (Q+U:  0.89287) (V:  0.0000) (T)
info string h3g2  (631 ) N:      21 (+ 0) (P:  1.61%) (Q:  0.00000) (D:  1.000) (U: 0.92416) (Q+U:  0.92416) (V:  0.0000) (T)
info string h3f1  (627 ) N:      22 (+ 0) (P:  1.69%) (Q:  0.00000) (D:  1.000) (U: 0.92711) (Q+U:  0.92711) (V:  0.0000) (T)
info string c4f1  (707 ) N:      23 (+ 0) (P:  1.73%) (Q:  0.00000) (D:  1.000) (U: 0.91256) (Q+U:  0.91256) (V:  0.0000) (T)
info string a2a1  (194 ) N:      24 (+ 0) (P:  1.78%) (Q:  0.00000) (D:  1.000) (U: 0.89724) (Q+U:  0.89724) (V:  0.0000) (T)
info string c4d3  (716 ) N:      24 (+ 0) (P:  1.80%) (Q:  0.00000) (D:  1.000) (U: 0.91111) (Q+U:  0.91111) (V:  0.0000) (T)
info string a2a3  (204 ) N:      26 (+ 0) (P:  1.95%) (Q:  0.00000) (D:  1.000) (U: 0.91283) (Q+U:  0.91283) (V:  0.0000) (T)
info string b3a1  (437 ) N:      26 (+ 0) (P:  1.98%) (Q:  0.00000) (D:  1.000) (U: 0.92388) (Q+U:  0.92388) (V:  0.0000) (T)
info string c3e2  (475 ) N:      27 (+ 0) (P:  2.02%) (Q:  0.00000) (D:  1.000) (U: 0.90877) (Q+U:  0.90877) (V:  0.0000) (T)
info string b3c1  (439 ) N:      31 (+ 0) (P:  2.30%) (Q:  0.00000) (D:  1.000) (U: 0.90804) (Q+U:  0.90804) (V:  0.0000) (T)
info string c4e2  (712 ) N:      31 (+ 0) (P:  2.35%) (Q:  0.00000) (D:  1.000) (U: 0.92701) (Q+U:  0.92701) (V:  0.0000) (T)
info string h3c8  (650 ) N:      33 (+ 0) (P:  2.45%) (Q:  0.00000) (D:  1.000) (U: 0.90930) (Q+U:  0.90930) (V:  0.0000) (T)
info string a2b1  (195 ) N:      38 (+ 0) (P:  2.85%) (Q:  0.00000) (D:  1.000) (U: 0.92336) (Q+U:  0.92336) (V:  0.0000) (T)
info string c3d1  (469 ) N:      62 (+ 0) (P:  4.65%) (Q:  0.00000) (D:  1.000) (U: 0.93102) (Q+U:  0.93102) (V:  0.0000) (T)
info string h3e6  (646 ) N:     117 (+ 0) (P:  8.68%) (Q:  0.00000) (D:  1.000) (U: 0.92819) (Q+U:  0.92819) (V:  0.0000) (T)
info string c3b1  (467 ) N:     145 (+ 0) (P: 10.75%) (Q:  0.00000) (D:  1.000) (U: 0.92935) (Q+U:  0.92935) (V:  0.0000) (T)
info string a2b2  (197 ) N:     146 (+ 0) (P: 10.86%) (Q:  0.00000) (D:  1.000) (U: 0.93220) (Q+U:  0.93220) (V:  0.0000) (T)
info string b3d2  (444 ) N:     155 (+ 0) (P: 11.53%) (Q:  0.00000) (D:  1.000) (U: 0.93324) (Q+U:  0.93324) (V:  0.0000) (T)
info string b3a5  (456 ) N:     379 (+ 0) (P:  1.26%) (Q:  0.88505) (D:  0.000) (U: 0.04186) (Q+U:  0.92691) (V:  0.9524)
info string b8b6  (1631) N:   49790 (+534) (P: 16.88%) (Q:  0.92491) (D:  0.009) (U: 0.00423) (Q+U:  0.92915) (V:  0.9700)
bestmove b8b6 ponder e7d8

```

As you can see, leela does not postpone anymore but goes for the b6-pawn capture immediately.
Also on much lower nodecounts, Rb8xb6 is played.
 
So, being able to adjust the draw-move-rule from 100 ply to a lower number can show interesting analyses results.

Happy analysing!

### CODE REVIEW

**Comment by Tilps** [comment]
I suggest you might find it more interesting to have a parameter that adjusts what the no capture no pawn ply counter is reset to when a capture or pawn move occurs.  By setting that to a number greater than 0 it simulates reduced 50 move rule while also asking the NN to evalutate like it is closer to a draw.

**Comment by MelleKoning** [comment]
Hi @Tilps updated the code so that no_capture_ply is initialized greather than zero. 
I've run the position_tests and they work fine; I do not understand why the Blas appveyor fails over that test.

**Comment by Naphthalin** [comment]
I think the intended behavior induced by this PR will be obsolete with the MLH implementation, and the underlying issue should be fixed during training, not during play. I therefore recommend to close this PR.

**Comment by MelleKoning** [comment]
This is an example of a PR that is implemented based on requests that come back now and then on discord.
As it should not harm the strength it just adds a feature for analysis. Therefore please consider to just merge the feature.

File: src/chess/position.cc [hunk 1]
```diff
@@ -30,14 +30,24 @@
...
+
 Position::Position(const Position& parent, Move m)
     : no_capture_ply_(parent.no_capture_ply_ + 1),
       ply_count_(parent.ply_count_ + 1) {
   them_board_ = parent.us_board_;
   const bool capture = them_board_.ApplyMove(m);
   us_board_ = them_board_;
   us_board_.Mirror();
-  if (capture) no_capture_ply_ = 0;
+  if (capture) no_capture_ply_ = 0 + (100 - drawmoverule_);
```
Comment (mooskagh) [inline src/chess/position.cc:50]: Why is this moved in such a weird way (setting draw to 20 no-capture-moves is done by setting no_capture_ply_ to 60 ply), instead of just changing it in `SearchWorker::ExtendNode`?

If you do that for NN to see high values when encoding the position, it's better to do that in position encoding function (as a separate or the same parameter) rather than changing Position class.

Comment (MelleKoning) [reply src/chess/position.cc:50]: I don't know anymore, this PR is from last year. I'll close it.

**Comment by MelleKoning** [comment]
too far away from master; closed.
